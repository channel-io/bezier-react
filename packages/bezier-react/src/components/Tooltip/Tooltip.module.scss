.Tooltip {
  z-index: var(--z-index-tooltip);

  box-sizing: border-box;
  width: max-content;
  max-width: 260px;
  height: max-content;
  padding: 5px 8px;
  word-break: normal;
  word-wrap: break-word;

  background-color: var(--bg-white-low);
  overflow: hidden;
  border-radius: var(--radius-8);
  box-shadow: var(--ev-3);
}

.TooltipContainer {
  overflow: hidden;
}

.TooltipContent {
  white-space: pre-wrap;
}

.Icon {
  padding: 1px;
}
