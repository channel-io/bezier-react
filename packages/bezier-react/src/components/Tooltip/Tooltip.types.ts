import { type BezierIcon } from '@channel.io/bezier-icons'

import {
  type BezierComponentProps,
  type ChildrenProps,
  type ContentProps,
  type DisableProps,
} from '~/src/types/ComponentProps'

/**
 * An enumeration that determines the position of `Tooltip`.
 */
export enum TooltipPosition {
  TopCenter = 'topCenter',
  TopLeft = 'topLeft',
  TopRight = 'topRight',
  RightCenter = 'rightCenter',
  RightTop = 'rightTop',
  RightBottom = 'rightBottom',
  BottomCenter = 'bottomCenter',
  BottomLeft = 'bottomLeft',
  BottomRight = 'bottomRight',
  LeftCenter = 'leftCenter',
  LeftTop = 'leftTop',
  LeftBottom = 'leftBottom',
}

interface TooltipProviderOptions {
  /**
   * Keeps the content of the tooltip open on hover. Disabling this feature affects accessibility.
   * @default false
   */
  allowHover?: boolean
  /**
   * The delay from when the mouse enters a tooltip trigger until the tooltip opens.
   * @default 300
   */
  delayShow?: number
  /**
   * The delay from when the mouse leaves a tooltip content until the tooltip hides.
   * @default 0
   */
  delayHide?: number
  /**
   * How much time a user has to enter another trigger without incurring a delay again.
   * @default 500
   */
  skipDelayShow?: number
}

interface TooltipOptions {
  /**
   * The open state of the tooltip when it is initially rendered.
   */
  defaultShow?: boolean
  /**
   * An element that sits below the tooltip content.
   */
  description?: React.ReactNode
  /**
   * A decorative icon that sits right of the tooltip content.
   */
  icon?: BezierIcon
  /**
   * Options to determine the location from the trigger.
   * @default TooltipPosition.BottomCenter
   */
  placement?: TooltipPosition
  /**
   * The distance in pixels from the trigger.
   * @default 4
   */
  offset?: number
  /**
   * Specify a container element to portal the content into.
   * @note When placed inside a `Modal`, default value is the container element of `Modal`.
   * @default document.body
   */
  container?: HTMLElement | null
  /**
   * When `true`, overrides the `position` of the tooltip
   * to prevent collisions with boundary edges.
   * @default true
   */
  keepInContainer?: boolean
  /**
   * Keeps the content of the tooltip open on hover. Disabling this feature affects accessibility.
   * Inherits from the nearest `TooltipProvider`.
   * @default false
   */
  allowHover?: boolean
  /**
   * The delay from when the mouse enters a tooltip trigger until the tooltip opens.
   * Inherits from the nearest `TooltipProvider`.
   * @default 300
   */
  delayShow?: number
  /**
   * The delay from when the mouse leaves a tooltip content until the tooltip hides.
   * Inherits from the nearest `TooltipProvider`.
   * @default 0
   */
  delayHide?: number
  /**
   * Callback function to be called when the tooltip is opened.
   */
  onShow?: () => void
  /**
   * Callback function to be called when the tooltip is closed.
   */
  onHide?: () => void
  /**
   * Event handler called when the escape key is down.
   * It can be prevented by calling `event.preventDefault`.
   */
  onEscapeKeyDown?: (event: KeyboardEvent) => void
  /**
   * Event handler called when a pointer event occurs outside the bounds of the component.
   * It can be prevented by calling `event.preventDefault`.
   */
  onPointerDownOutside?: (event: CustomEvent<{ originalEvent: PointerEvent }>) => void
}

export interface TooltipProviderProps extends
  ChildrenProps,
  TooltipProviderOptions {}

export interface TooltipImplProps extends
  BezierComponentProps,
  ChildrenProps<React.ReactElement>,
  ContentProps,
  Omit<React.HTMLAttributes<HTMLDivElement>, keyof ContentProps | keyof ChildrenProps>,
  TooltipOptions {}

export interface TooltipProps extends
  DisableProps,
  TooltipImplProps {}
