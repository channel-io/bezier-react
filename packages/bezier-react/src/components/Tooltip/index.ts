export {
  Tooltip,
  TooltipProvider,
} from './Tooltip'

export { TooltipPosition } from './Tooltip.types'

export type {
  TooltipProps,
  TooltipProviderProps,
} from './Tooltip.types'
