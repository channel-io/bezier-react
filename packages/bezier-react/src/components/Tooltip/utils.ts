/* Internal dependencies */
import { getRootElement } from '~/src/utils/domUtils'
import {
  TooltipPosition,
  type GetTooltipStyle,
  type GetReplacement,
} from './Tooltip.types'

export function getReplacement({
  tooltip,
  keepInContainer,
  placement,
}: GetReplacement): TooltipPosition {
  if (!keepInContainer) {
    return placement
  }

  const {
    width: tooltipWidth,
    height: tooltipHeight,
    top: tooltipTop,
    left: tooltipLeft,
  } = tooltip.getBoundingClientRect()
  const {
    width: rootWidth,
    height: rootHeight,
    top: rootTop,
    left: rootLeft,
  } = getRootElement().getBoundingClientRect()

  const isOverTop = tooltipTop < rootTop
  const isOverBottom = tooltipTop + tooltipHeight > rootTop + rootHeight
  const isOverLeft = tooltipLeft < rootLeft
  const isOverRight = tooltipLeft + tooltipWidth > rootLeft + rootWidth

  if (isOverTop) {
    if (isOverLeft) {
      return TooltipPosition.BottomLeft
    }
    if (isOverRight) {
      return TooltipPosition.BottomRight
    }
    return TooltipPosition.BottomCenter
  }

  if (isOverBottom) {
    if (isOverLeft) {
      return TooltipPosition.TopLeft
    }
    if (isOverRight) {
      return TooltipPosition.TopRight
    }
    return TooltipPosition.TopCenter
  }

  if (isOverLeft) {
    if (isOverTop) {
      return TooltipPosition.RightTop
    }
    if (isOverBottom) {
      return TooltipPosition.RightBottom
    }
    return TooltipPosition.RightCenter
  }

  if (isOverRight) {
    if (isOverTop) {
      return TooltipPosition.LeftTop
    }
    if (isOverBottom) {
      return TooltipPosition.LeftBottom
    }
    return TooltipPosition.LeftCenter
  }

  return placement
}

export function getTooltipStyle({
  tooltipContainer,
  placement,
  offset,
  allowHover,
}: GetTooltipStyle) {
  const {
    top: containerTop,
    left: containerLeft,
    width: containerWidth,
    height: containerHeight,
  } = tooltipContainer.getBoundingClientRect()

  let top = containerTop
  let left = containerLeft
  let translateX = 0
  let translateY = 0

  let offsetTop = 0
  let offsetRight = 0
  let offsetBottom = 0
  let offsetLeft = 0
  let offsetStyle = {}

  switch (placement) {
    case TooltipPosition.TopCenter:
    case TooltipPosition.TopLeft:
    case TooltipPosition.TopRight:
      translateY = -100
      offsetBottom = offset
      break
    case TooltipPosition.RightCenter:
    case TooltipPosition.RightTop:
    case TooltipPosition.RightBottom:
      left += containerWidth
      offsetLeft = offset
      break
    case TooltipPosition.BottomCenter:
    case TooltipPosition.BottomLeft:
    case TooltipPosition.BottomRight:
      top += containerHeight
      offsetTop = offset
      break
    case TooltipPosition.LeftCenter:
    case TooltipPosition.LeftTop:
    case TooltipPosition.LeftBottom:
      translateX = -100
      offsetRight = offset
      break
  }

  switch (placement) {
    case TooltipPosition.TopCenter:
    case TooltipPosition.BottomCenter:
      translateX = -50
      left += (containerWidth / 2)
      break
    case TooltipPosition.TopRight:
    case TooltipPosition.BottomRight:
      translateX = -100
      left += containerWidth
      break
    case TooltipPosition.RightCenter:
    case TooltipPosition.LeftCenter:
      translateY = -50
      top += (containerHeight / 2)
      break
    case TooltipPosition.RightBottom:
    case TooltipPosition.LeftBottom:
      translateY = -100
      top += containerHeight
      break
  }

  if (allowHover) {
    offsetStyle = {
      paddingTop: offsetTop,
      paddingRight: offsetRight,
      paddingBottom: offsetBottom,
      paddingLeft: offsetLeft,
    }
  } else {
    top += (offsetTop - offsetBottom)
    left += (offsetLeft - offsetRight)
  }

  return {
    position: 'fixed',
    top,
    left,
    transform: `translate(${translateX}%, ${translateY}%)`,
    ...offsetStyle,
  }
}
