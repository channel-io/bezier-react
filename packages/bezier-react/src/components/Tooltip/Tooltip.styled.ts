import {
  LineHeightAbsoluteNumber,
  Typography,
  ellipsis,
  styled,
} from '~/src/foundation'

import { ZIndex } from '~/src/constants/ZIndex'

import { AlphaStack } from '~/src/components/AlphaStack'
import {
  Icon as BaseIcon,
  IconSize,
} from '~/src/components/Icon'
import { Text } from '~/src/components/Text'

import { type TooltipProps } from './Tooltip.types'

export const TooltipContent = styled(AlphaStack).attrs({
  direction: 'horizontal',
  spacing: 4,
})<TooltipProps>`
  z-index: ${ZIndex.Tooltip};

  box-sizing: border-box;
  width: max-content;
  max-width: 260px;
  height: max-content;
  padding: 5px 8px;
  word-break: normal;
  word-wrap: break-word;

  ${({ foundation }) => foundation?.elevation?.ev3(true)};
  ${({ foundation }) => foundation?.rounding?.round8};

  ${({ interpolation }) => interpolation}
`

export const TextContainer = styled.div`
  overflow: hidden;
`

export const Title = styled(Text).attrs({ typo: Typography.Size13, bold: true })`
  margin-bottom: 2px;
  color: ${({ foundation }) => foundation?.subTheme?.['txt-black-darkest']};
`

export const Content = styled(Text).attrs({ typo: Typography.Size13 })`
  color: ${({ foundation }) => foundation?.subTheme?.['txt-black-darkest']};
  /* NOTE: Line height of Typography.Size13  */
  ${ellipsis(20, LineHeightAbsoluteNumber.Lh18)}
  white-space: pre-wrap;
`

export const Description = styled(Text).attrs({ typo: Typography.Size12 })`
  color: ${({ foundation }) => foundation?.subTheme?.['txt-black-dark']};
`

export const Icon = styled(BaseIcon).attrs({
  size: IconSize.XS,
})`
  padding: 1px;
  color: ${({ foundation }) => foundation?.subTheme?.['txt-black-darkest']};
`
