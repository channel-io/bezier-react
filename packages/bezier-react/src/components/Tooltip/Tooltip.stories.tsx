/* External dependencies */
import React, { useState } from 'react'
import { base } from 'paths.macro'
import {
  type Story,
  type Meta,
} from '@storybook/react'

/* Internal dependencies */
import { styled } from '~/src/foundation'
import { getTitle } from '~/src/utils/storyUtils'
import Tooltip from './Tooltip'
import type TooltipProps from './Tooltip.types'
import { TooltipPosition } from './Tooltip.types'

export default {
  title: getTitle(base),
  component: Tooltip,
  argTypes: {
    content: {
      control: {
        type: 'text',
      },
    },
    placement: {
      control: {
        type: 'radio',
        options: [
          TooltipPosition.TopCenter,
          TooltipPosition.TopLeft,
          TooltipPosition.TopRight,
          TooltipPosition.RightCenter,
          TooltipPosition.RightTop,
          TooltipPosition.RightBottom,
          TooltipPosition.BottomCenter,
          TooltipPosition.BottomLeft,
          TooltipPosition.BottomRight,
          TooltipPosition.LeftCenter,
          TooltipPosition.LeftTop,
          TooltipPosition.LeftBottom,
        ],
      },
    },
    offset: {
      control: {
        type: 'range',
        min: 0,
        max: 50,
        step: 1,
      },
    },
    delayShow: {
      control: {
        type: 'range',
        min: 0,
        max: 5000,
        step: 100,
      },
    },
    delayHide: {
      control: {
        type: 'range',
        min: 0,
        max: 5000,
        step: 100,
      },
    },
  },
} as Meta

const Target = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100px;
  height: 40px;
  background-color: ${props => props.foundation?.theme?.['bg-black-dark']};
  border-radius: 4px;
`

const Template: Story<TooltipProps> = (props) => {
  const [container, setContainer] = useState<HTMLDivElement | null>(null)

  return (
    <div ref={setContainer}>
      <Tooltip
        {...props}
        container={container}
      >
        <Target>
          Target
        </Target>
      </Tooltip>
    </div>
  )
}

export const Primary = Template.bind({})

Primary.args = {
  // eslint-disable-next-line max-len
  content: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry\'s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.',
  lazy: true,
  placement: TooltipPosition.BottomCenter,
  offset: 4,
  disabled: false,
  keepInContainer: false,
  allowHover: false,
  delayShow: 0,
  delayHide: 0,
}
