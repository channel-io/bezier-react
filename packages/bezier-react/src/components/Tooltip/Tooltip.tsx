import React, {
  forwardRef,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react'

import * as TooltipPrimitive from '@radix-ui/react-tooltip'

import { document } from '~/src/utils/domUtils'
import { isBoolean } from '~/src/utils/typeUtils'

import {
  TooltipPosition,
  type TooltipProps,
  type TooltipProviderProps,
} from './Tooltip.types'

import * as Styled from './Tooltip.styled'

// TODO: (@ed) Evolve it into a commonly reusable function
// FIXME: duplicate
function createContext<ContextValue>(
  providerName: string,
  defaultValue: ContextValue,
) {
  const Context = React.createContext<ContextValue>(defaultValue)

  function useContext(consumerName: string) {
    const contextValue = React.useContext(Context)

    if (!contextValue) {
      throw new Error(`'${consumerName}' must be used within '${providerName}'`)
    }

    return contextValue
  }

  return [
    Context.Provider,
    useContext,
  ] as const
}

function getSideAndAlign(
  placement: TooltipPosition,
): Pick<TooltipPrimitive.TooltipContentProps, 'side' | 'align'> {
  switch (placement) {
    case TooltipPosition.TopCenter:
      return {
        side: 'top',
        align: 'center',
      }
    case TooltipPosition.TopLeft:
      return {
        side: 'top',
        align: 'start',
      }
    case TooltipPosition.TopRight:
      return {
        side: 'top',
        align: 'end',
      }
    case TooltipPosition.RightCenter:
      return {
        side: 'right',
        align: 'center',
      }
    case TooltipPosition.RightTop:
      return {
        side: 'right',
        align: 'start',
      }
    case TooltipPosition.RightBottom:
      return {
        side: 'right',
        align: 'end',
      }
    case TooltipPosition.BottomCenter:
      return {
        side: 'bottom',
        align: 'center',
      }
    case TooltipPosition.BottomLeft:
      return {
        side: 'bottom',
        align: 'start',
      }
    case TooltipPosition.BottomRight:
      return {
        side: 'bottom',
        align: 'end',
      }
    case TooltipPosition.LeftCenter:
      return {
        side: 'left',
        align: 'center',
      }
    case TooltipPosition.LeftTop:
      return {
        side: 'left',
        align: 'start',
      }
    case TooltipPosition.LeftBottom:
      return {
        side: 'left',
        align: 'end',
      }
    default:
      // NOTE: should not reach here
      return {
        side: undefined,
        align: undefined,
      }
  }
}

const [
  /**
   * NOTE: Custom context use because the radix-ui doesn't support `delayHide`.
   */
  TooltipGlobalContextProvider,
  useTooltipGlobalContext,
] = createContext<Required<Pick<TooltipProviderProps, 'delayHide'>> | null>('TooltipProvider', null)

/**
 * `TooltipProvider` is used to globally provide props to child tooltips.
 *
 * @example
 *
 * ```tsx
 * <TooltipProvider allowHover delayShow={1000}>
 *   <Tooltip />
 * </TooltipProvider>
 * ```
 */
export function TooltipProvider({
  children,
  allowHover = false,
  delayShow = 0,
  delayHide = 0,
  skipDelayShow = 0,
}: TooltipProviderProps) {
  const contextValue = useMemo(() => ({
    delayHide,
  }), [delayHide])

  return (
    <TooltipPrimitive.Provider
      delayDuration={delayShow}
      skipDelayDuration={skipDelayShow}
      disableHoverableContent={!allowHover}
    >
      <TooltipGlobalContextProvider value={contextValue}>
        { children }
      </TooltipGlobalContextProvider>
    </TooltipPrimitive.Provider>
  )
}

/**
 * `Tooltip` is a component that shows additional information when the mouse hovers or the keyboard is focused.
 *
 * @example
 *
 * ```tsx
 * // Anatomy of the Tooltip
 * <TooltipProvider>
 *   <Tooltip />
 * </TooltipProvider>
 *
 * // Example of a Tooltip with a button
 * <Tooltip content="Ta-da!">
 *   <button>Hover me</button>
 * </Tooltip>
 * ```
 */
export const Tooltip = forwardRef<HTMLDivElement, TooltipProps>(function Tooltip({
  as,
  children,
  defaultShow,
  onShow: onShowProp,
  onHide: onHideProp,
  disabled,
  content,
  description,
  icon,
  placement = TooltipPosition.BottomCenter,
  offset = 4,
  container = document.body,
  keepInContainer = true,
  allowHover,
  delayShow,
  delayHide: delayHideProp,
  ...rest
}, forwardedRef) {
  const [show, setShow] = useState<boolean>(defaultShow ?? false)
  const timeoutRef = useRef<NodeJS.Timeout>()
  const { delayHide: globalDelayHide } = useTooltipGlobalContext('Tooltip')

  const delayHide = delayHideProp ?? globalDelayHide

  useEffect(function cleanUpTimeout() {
    return function cleanUp() {
      if (timeoutRef.current) {
        clearTimeout(timeoutRef.current)
      }
    }
  }, [])

  const onShow = useCallback(() => {
    setShow(true)
    onShowProp?.()
  }, [onShowProp])

  const onHide = useCallback(() => {
    setShow(false)
    onHideProp?.()
  }, [onHideProp])

  const onOpenChange = useCallback((open: boolean) => {
    if (disabled) { return }

    if (open) {
      onShow()
      return
    }

    if (timeoutRef.current) {
      clearTimeout(timeoutRef.current)
      timeoutRef.current = undefined
    }

    if (delayHide > 0) {
      timeoutRef.current = setTimeout(() => {
        onHide()
      }, delayHide)
      return
    }

    onHide()
  }, [
    disabled,
    delayHide,
    onShow,
    onHide,
  ])

  return (
    <TooltipPrimitive.Root
      open={show}
      defaultOpen={defaultShow}
      delayDuration={delayShow}
      disableHoverableContent={isBoolean(allowHover) ? !allowHover : undefined}
      onOpenChange={onOpenChange}
    >
      <TooltipPrimitive.Trigger asChild>
        { children }
      </TooltipPrimitive.Trigger>

      <TooltipPrimitive.Portal container={container}>
        <TooltipPrimitive.Content
          {...rest}
          {...getSideAndAlign(placement)}
          asChild
          ref={forwardedRef}
          sideOffset={offset}
          avoidCollisions={keepInContainer}
          collisionPadding={8}
          hideWhenDetached
        >
          <Styled.TooltipContent forwardedAs={as}>
            <div>
              <Styled.Content>
                { content }
              </Styled.Content>

              { description && (
                <Styled.Description>
                  { description }
                </Styled.Description>
              ) }
            </div>

            { icon && (
              <Styled.Icon source={icon} />
            ) }
          </Styled.TooltipContent>
        </TooltipPrimitive.Content>
      </TooltipPrimitive.Portal>
    </TooltipPrimitive.Root>
  )
})
