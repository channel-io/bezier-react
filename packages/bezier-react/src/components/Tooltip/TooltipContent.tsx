/* External dependencies */
import React, {
  useState,
  useCallback,
  useMemo,
  useRef,
  useEffect,
} from 'react'
import ReactDOM from 'react-dom'

/* Internal dependencies */
import { Typography } from '~/src/foundation'
import useMergeRefs from '~/src/hooks/useMergeRefs'
import useEventHandler from '~/src/hooks/useEventHandler'
import {
  isArray,
  isString,
  isEmpty,
} from '~/src/utils/typeUtils'
import { getRootElement } from '~/src/utils/domUtils'
import { Text } from '~/src/components/Text'
import {
  type TooltipContentProps,
  TooltipPosition,
} from './Tooltip.types'
import {
  getReplacement,
  getTooltipStyle,
} from './utils'
import {
  ContentWrapper,
  Content,
  EllipsisableContent,
} from './Tooltip.styled'

function getNewLineComponent(strContent: string) {
  return (
    strContent.split('\n').map((str, index) => {
      if (index === 0) {
        return (
          <Text key={str} typo={Typography.Size14}>
            { str }
          </Text>
        )
      }

      return (
        <React.Fragment key={str}>
          <br />
          <Text typo={Typography.Size14}>
            { str }
          </Text>
        </React.Fragment>
      )
    })
  )
}

function getContentComponent(content?: React.ReactNode) {
  if (isArray(content)) {
    return content.map(item => {
      if (isString(item)) {
        return getNewLineComponent(item)
      }

      return item
    })
  }

  if (isString(content)) {
    return getNewLineComponent(content)
  }

  return content
}

const TooltipContent: React.FC<TooltipContentProps> = ({
  as,
  content,
  contentClassName,
  contentInterpolation,
  disabled = false,
  container: givenContainer,
  keepInContainer = false,
  placement = TooltipPosition.BottomCenter,
  tooltipContainer,
  offset = 4,
  allowHover = false,
  testId,
  forwardedRef,
}) => {
  const tooltipRef = useRef<HTMLDivElement>(null)
  const tooltipWrapperRef = useRef<HTMLDivElement>(null)
  const mergedRef = useMergeRefs<HTMLDivElement>(tooltipRef, forwardedRef)
  const [replacement, setReplacement] = useState(placement)

  const container = givenContainer || getRootElement()

  const handleClickTooltip = useCallback((event: HTMLElementEventMap['click']) => {
    event.stopPropagation()
  }, [])

  useEventHandler(tooltipRef.current, 'click', handleClickTooltip)

  useEffect(() => {
    if (!tooltipRef.current) { return }
    const newPlacement = getReplacement({
      tooltip: tooltipRef.current,
      container,
      keepInContainer,
      placement,
    })
    setReplacement(newPlacement)
  }, [
    container,
    keepInContainer,
    placement,
  ])

  const ContentComponent = useMemo(() => getContentComponent(content), [content])

  const contentWrapperStyle = useMemo(() => {
    if (tooltipContainer) {
      return getTooltipStyle({
        tooltipContainer,
        placement: replacement,
        offset,
        allowHover,
      })
    }

    return {}
  }, [
    tooltipContainer,
    replacement,
    offset,
    allowHover,
  ])

  return (
    ReactDOM.createPortal(
      <ContentWrapper
        ref={tooltipWrapperRef}
        disabled={disabled || isEmpty(content)}
        style={contentWrapperStyle}
      >
        <Content
          as={as}
          data-testid={testId}
          className={contentClassName}
          interpolation={contentInterpolation}
          ref={mergedRef}
        >
          <EllipsisableContent>
            { ContentComponent }
          </EllipsisableContent>
        </Content>
      </ContentWrapper>,
      container,
    )
  )
}

export default TooltipContent
