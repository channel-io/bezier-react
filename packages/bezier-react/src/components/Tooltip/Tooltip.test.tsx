/* External dependencies */
import React from 'react'
import { act } from '@testing-library/react'
import { fireEvent } from '@testing-library/dom'

/* Internal dependencies */
import { css } from 'Foundation'
import { render } from 'Utils/testUtils'
import Tooltip, { TOOLTIP_TEST_ID, TOOLTIP_CONTENT_TEST_ID } from './Tooltip'
import TooltipProps from './Tooltip.types'

const RootTooltip: React.FC<TooltipProps> = ({ children, ...rests }) => (
  <Tooltip {...rests}>
    { children }
  </Tooltip>
)

describe('Tooltip test >', () => {
  let props: TooltipProps

  const content = 'Hovered'

  beforeEach(() => {
    props = {
      children: 'Text',
      content,
    }

    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.useRealTimers()
  })

  const renderTooltip = (optionProps?: TooltipProps) => render(
    <RootTooltip {...props} {...optionProps} />,
  )

  it('Tooltip with default props', () => {
    const { baseElement, getByTestId } = renderTooltip()

    act(() => {
      fireEvent.mouseOver(getByTestId(TOOLTIP_TEST_ID))

      jest.runAllTimers()
    })

    expect(baseElement).toMatchSnapshot()
  })

  it('Tooltip with contentInterpolation prop', async () => {
    const { baseElement, getByTestId } = renderTooltip({
      contentInterpolation: css`background-color: black;`,
    })

    act(() => {
      fireEvent.mouseOver(getByTestId(TOOLTIP_TEST_ID))

      jest.runAllTimers()
    })

    expect(baseElement).toMatchSnapshot()
  })

  it('TooltipContent not rendered at first', async () => {
    const { queryByTestId } = renderTooltip()

    expect(queryByTestId(TOOLTIP_CONTENT_TEST_ID)).toBeNull()
  })

  it('TooltipContent rendered after mouseover', async () => {
    const { getByTestId, queryByTestId } = renderTooltip()

    act(() => {
      fireEvent.mouseOver(getByTestId(TOOLTIP_TEST_ID))

      jest.runAllTimers()
    })

    expect(queryByTestId(TOOLTIP_CONTENT_TEST_ID)).toBeVisible()
  })

  it('Tooltip with disabled prop not rendered even after mouseover', async () => {
    const { getByTestId, queryByTestId } = renderTooltip({ disabled: true })

    act(() => {
      fireEvent.mouseOver(getByTestId(TOOLTIP_TEST_ID))

      jest.runAllTimers()
    })

    expect(queryByTestId(TOOLTIP_CONTENT_TEST_ID)).toBeNull()
  })
})
