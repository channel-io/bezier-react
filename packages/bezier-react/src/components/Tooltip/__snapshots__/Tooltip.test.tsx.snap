// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip test > Tooltip with contentInterpolation prop 1`] = `
.c4 {
  font-size: 1.4rem;
  line-height: 1.8rem;
  margin: 0px 0px 0px 0px;
  font-style: normal;
  font-weight: normal;
  color: inherit;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c0 {
  position: relative;
}

.c1 {
  position: absolute;
  top: 0;
  left: 0;
  z-index: 1000000000;
}

.c2 {
  box-sizing: border-box;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  max-width: 260px;
  height: -webkit-max-content;
  height: -moz-max-content;
  height: max-content;
  padding: 8px 14px;
  color: #FFFFFFCC;
  word-break: normal;
  word-wrap: break-word;
  background-color: #313234;
  box-shadow: inset 0 0 2px 0 #FFFFFF14, 0 0 2px 1px #00000014, 0 2px 6px #00000026;
  overflow: hidden;
  border-radius: 8px;
  background-color: black;
}

.c3 {
  display: -webkit-box;
  max-height: 36rem;
  overflow: hidden;
  line-height: 1.8rem;
  text-overflow: ellipsis;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 20;
}

<body>
  <div>
    <div
      class="c0"
      data-testid="bezier-react-tooltip"
    >
      Text
    </div>
  </div>
  <div
    class="c1"
    style="position: fixed; top: 4px; left: 0px; transform: translate(-50%, 0%);"
  >
    <div
      class="c2"
      data-testid="bezier-react-tooltip-content"
    >
      <div
        class="c3"
      >
        <span
          class="c4"
          data-testid="bezier-react-text"
        >
          Hovered
        </span>
      </div>
    </div>
  </div>
</body>
`;

exports[`Tooltip test > Tooltip with default props 1`] = `
.c4 {
  font-size: 1.4rem;
  line-height: 1.8rem;
  margin: 0px 0px 0px 0px;
  font-style: normal;
  font-weight: normal;
  color: inherit;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c0 {
  position: relative;
}

.c1 {
  position: absolute;
  top: 0;
  left: 0;
  z-index: 1000000000;
}

.c2 {
  box-sizing: border-box;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  max-width: 260px;
  height: -webkit-max-content;
  height: -moz-max-content;
  height: max-content;
  padding: 8px 14px;
  color: #FFFFFFCC;
  word-break: normal;
  word-wrap: break-word;
  background-color: #313234;
  box-shadow: inset 0 0 2px 0 #FFFFFF14, 0 0 2px 1px #00000014, 0 2px 6px #00000026;
  overflow: hidden;
  border-radius: 8px;
}

.c3 {
  display: -webkit-box;
  max-height: 36rem;
  overflow: hidden;
  line-height: 1.8rem;
  text-overflow: ellipsis;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 20;
}

<body>
  <div>
    <div
      class="c0"
      data-testid="bezier-react-tooltip"
    >
      Text
    </div>
  </div>
  <div
    class="c1"
    style="position: fixed; top: 4px; left: 0px; transform: translate(-50%, 0%);"
  >
    <div
      class="c2"
      data-testid="bezier-react-tooltip-content"
    >
      <div
        class="c3"
      >
        <span
          class="c4"
          data-testid="bezier-react-text"
        >
          Hovered
        </span>
      </div>
    </div>
  </div>
</body>
`;
